cpp/src/arrow/array.cc

```diff
@@ -393,6 +393,71 @@ UnionArray::UnionArray(const std::shared_ptr<DataType>& type, int64_t length,
   SetData(internal_data);
 }
 
+std::shared_ptr<DataType> MakeUnionArrayType(
+    UnionMode mode, const std::vector<std::shared_ptr<Array>>& children) {
+  auto types = std::vector<std::shared_ptr<Field>>();
```

Comment:
This might be better with the type on the LHS (`std::vector<std::shared_ptr<Field>> types;`)